dash.el

```diff
@@ -1661,7 +1661,7 @@ SOURCE is a proper or improper list."
        ((cdr match-form)
         (cond
          ((and (symbolp (car match-form))
-               (memq (car match-form) '(&keys &plist &alist &hash)))
+               (functionp (intern (format "dash-expand:%s" (car match-form)))))
           (dash--match-kv (dash--match-kv-normalize-match-form match-form) (dash--match-cons-get-cdr skip-cdr source)))
```

Comment:
Does it make sense to call this `match-kv` now? That stands for key-value but clearly this custom thing can match anything.

There is also the call to `dash--match-kv-normalize-match-form` which for most other things makes no sense as it generates automatic places from key names.  If you have no keys then this will generate nonsense.

I think the check should go to a new branch.

Reply:
I am not sure I understand the comment. Are you suggesting to rename match-kv related internal functions to something else?

As a side note, I found out that the elisp linter fails to handle the newly added destructorings. I am not sure how to fix that, I will try to make the extension functions `autoload`-able.

Reply:
Ignore the last part, the problem with the checker what that it was using the `dash` that was downloaded from melpa, not the one that I have cloned.

Reply:
I'm suggesting to create a new branch in the cond for the additional check and leave the kv stuff as it is because it has special handling for some kv related things which might not be applicable in general for any matcher that the user adds.

Reply:
Are you are referring the `smarter` destructoring like `(&hash "key1" "key2")`? If this is the case IMO it would be better to keep it as in the PR in order to share that logic. If there is a need for destructoring for which the kv related stuff is not applicable there could be another extension point.

Reply:
But this patch does not "extend kv" but anything.

Imagine you write a matcher for a list, writing `(&list a b)` would instead generate `(%list 'a a 'b b)` with the current code.  That makes no sense.

Reply:
Or maybe it would just flat-out error because that can't in fact be expanded.

Reply:
This PR targets only generation of KV templates -  it does not handle all of the cases and I haven't had that as a goal(although I cannot think of a pattern to which cannot be handled). I have already commented what are the benefits from sharing the kv related logic and you could see that in the added tests. If you do not like the PR and you think that does not bring value feel free to reject it. IMO it could be even viewed as a refactoring since it removed the &hash/&key/&alist/&plist usages through the code.

Furthermore, even your example `(&list a b)` will work if the KV logic is changed to use keys `0` and `1`...

Reply:
I'm sorry but I don't understand.  You opened the issue https://github.com/magnars/dash.el/issues/276 and this is supposed to solve that.  But it does something else.  That's where my confusion is coming from.

Was the goal here not to address #276?

Reply:
Here it is the description of the issue...

> At this point, `dash` supports 4 different destructoring keyword (keys/hash/plist/alist) and from what I can see it will be not hard to add new keywords by providing a destructuring template(or macro?) to generate the getter.
> 
> Example usecases:
> 
>     * `&hash` which checks whether the map is empty before applying gethash
> 
>     * Custom destructuring for `cl-defstruct`, e. g.
> 
> 
> ```emacs-lisp
> (-let [(&my/struct slot1 slot2) value]
>   ...)
> ```

... and I did exactly what I have described covering the usecases I have mentioned. Note that I have described the implementation details so the maintainers could comment upfront whether this approach is acceptable in order to avoid wasting my time doing a PR that wont be accepted.

Reply:
The first part at no place mentions key/value destructuring, only that we have 4 keywords now (which incidentally manage kv storage).

Structures could be solved in some generic way out of the box instead of requiring the user to add a destructuring definition for every structure.

The `&hash?` example is interesting.  Maybe we could automatically extend every keyword with `?` to check for the condition somehow.

The point is that if we give the user ability to extend this in arbitrary ways we should make sure it works as expected.  I already misunderstood this feature and I'm quite familiar with the whole thing.  So I imagine other people can get confused as well.

I'll need to play with this a bit more.

I am not against adding support for custom expanders, but it needs to make sense.  I don't want to end up with another pcase.

Reply:
> I am not against adding support for custom expanders, but it needs to make sense. I don't want to end up with another pcase.

FWIW, one person's `pcase` is another person's `pcase`. ;)

Reply:
> The first part at no place mentions key/value destructuring, only that we have 4 keywords now (which incidentally manage kv storage).

It is not incidentally, but there is no point to discuss that...

> Structures could be solved in some generic way out of the box instead of requiring the user to add a destructuring definition for every structure.

Yes they can, but it would require the name of the structure to make that work: Compare:

``` emacs-lisp
(-let (((&lsp-wk 'proc (&lsp-cl 'server-id)) workspace)
   ...)
```
with...
``` emacs-lisp
(-let (((&slots lsp--workspace-parser proc (&slots lsp--client server-id)) workspace)
   ...)
```

>
> The `&hash?` example is interesting. Maybe we could automatically extend every keyword with `?` to check for the condition somehow.

No, since it makes sense only for &hash. &hash? could go as part of the `dash` core.

> The point is that if we give the user ability to extend this in arbitrary ways we should make sure it works as expected. I already misunderstood this feature and I'm quite familiar with the whole thing. So I imagine other people can get confused as well.

I am comfused even more by your comments - in previous comment you said that this destructuring is NOT powerful enough but now you wonder whether to give that power to the users...

> I'll need to play with this a bit more.
>
> I am not against adding support for custom expanders, but it needs to make sense. I don't want to end up with another pcase.

Sure, take your time. `pcase` issues could be avoided by having namespace prefix to each destructoring key.